strawberryfields/device.py

```diff
@@ -142,18 +142,35 @@ def validate_parameters(self, **parameters: complex) -> None:
         if self.gate_parameters is None:
             return
 
+        def _flatten(iterable: Iterable) -> Generator:
+            for it in iterable:
+                if isinstance(it, Iterable) and not isinstance(it, str):
+                    yield from _flatten(it)
+                else:
+                    yield it
+
         # check that all provided parameters are valid
         for p, v in parameters.items():
-            if p in self.gate_parameters and v not in self.gate_parameters[p]:
-                # parameter is present in the device specifications
-                # but the user has provided an invalid value
-                raise ValueError(
-                    f"{p} has invalid value {v}. Only {self.gate_parameters[p]} allowed."
-                )
-
-            if p not in self.gate_parameters:
-                raise ValueError(f"Parameter {p} not a valid parameter for this device")
-
+            if p in self.gate_parameters:
+                if isinstance(v, Iterable):
+                    for i in _flatten(v):
+                        if i not in self.gate_parameters[p]:
+                            raise ValueError(
+                                f"'{p}' has invalid value {i}. Only {self.gate_parameters[p]} allowed."
+                            )
+                else:
+                    if v not in self.gate_parameters[p]:
+                        # parameter is present in the device specifications
+                        # but the user has provided an invalid value
+                        raise ValueError(
+                            f"'{p}' has invalid value {v}. Only {self.gate_parameters[p]} allowed."
+                        )
+
+            else:
+                raise ValueError(f"Parameter '{p}' not a valid parameter for this device")
```

Comment:
A small cosmetic suggestion: I think the code is more readable if the first `if-else` clause is replaced by a guard clause (halting the execution as well when the condition of `p in self.gate_parameters` is not met) and hence removing one level of nesting. 💅

```suggestion
            if p not in self.gate_parameters:
                raise ValueError(f"Parameter '{p}' not a valid parameter for this device")
            
            if isinstance(v, Iterable):
                for i in _flatten(v):
                    if i not in self.gate_parameters[p]:
                        raise ValueError(
                            f"'{p}' has invalid value {i}. Only {self.gate_parameters[p]} allowed."
                        )
            elif v not in self.gate_parameters[p]:
                # parameter is present in the device specifications
                # but the user has provided an invalid value
                raise ValueError(
                    f"'{p}' has invalid value {v}. Only {self.gate_parameters[p]} allowed."
                )
```

Reply:
Thanks, very good suggestion! 💯